### PR TITLE
Unit Tests for wrist

### PR DESCRIPTION
**Pull request by thobotics**
New unit tests for wrist.
Fix matrix explosion problem in forward_kinematics function.

### CODE REVIEW

File: scripts/wrist_test.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+
+    rospy.loginfo("Hard reset")
+    rospy.sleep(2.0)
+
+
+def soft_reset():
+    pass
+
+
+def hand_shake(lr="right"):
```
Comment (missxa) [inline scripts/wrist_test.py]: what's 'lr'?

File: scripts/wrist_test.py [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+    msg.position = [0]*len(msg.name)
+    msg.velocity = [0]*len(msg.name)
+    msg.effort = [0]*len(msg.name)
+    publisher.publish(msg)
+
+    rospy.loginfo("Hard reset")
+    rospy.sleep(2.0)
+
+
+def soft_reset():
```
Comment (missxa) [inline scripts/wrist_test.py]: why is it here if it's no implemented?

File: scripts/wrist_test.py [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+    pass
+
+
+def hand_shake(lr="right"):
+    msg = JointState()
+    msg.name = ["elbow_"+lr+"_axis0"]
+    msg.velocity = [0]*len(msg.name)
+    msg.effort = [0]*len(msg.name)
+    ranges = [0.6, 1.0]
```
Comment (missxa) [inline scripts/wrist_test.py]: pass ranges as function parameters

File: scripts/wrist_test.py [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+    samples = 50
+    for i in range(10):
+        if not i % 2:
+            rospy.loginfo("up")
```
Comment (missxa) [inline scripts/wrist_test.py]: add meaningful logging

File: scripts/wrist_test.py [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+                msg.velocity = [0]
+                msg.effort = [0]
+                publisher.publish(msg)
+                rate.sleep()
+                # rospy.sleep(0.1)
+            # rospy.sleep(2.0)
+            # msg.position = [-0.3]
+
+
+def move_shoulder_elbow(lr="right", elbow=True):
```
Comment (missxa) [inline scripts/wrist_test.py]: what are the parameters?

File: scripts/wrist_test.py [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+        msg.velocity = [0]*len(msg.name)
+        msg.effort = [0]*len(msg.name)
+        publisher.publish(msg)
+        rate.sleep()
+
+
+def move_wrist(lr="right", axis=0):
+    msg = JointState()
+
+    ranges = [-1.5, -0.05]
```
Comment (missxa) [inline scripts/wrist_test.py]: take ranges as a param

File: scripts/wrist_test.py [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+    samples = 200
```
Comment (missxa) [inline scripts/wrist_test.py]: this should be a param too

File: scripts/wrist_test.py [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+    msg_name = ["shoulder_left_axis0", "elbow_left_axis0",  "wrist_left_axis0", "wrist_left_axis1",
+                "shoulder_right_axis0", "elbow_right_axis0",  "wrist_right_axis0", "wrist_right_axis1",
+                "head_axis0"
+                ]
+    # ["shoulder_"+lr+"_axis0", "elbow_"+lr+"_axis0",  "wrist_"+lr+"_axis0", "wrist_"+lr+"_axis1", "head_axis0"]
+    msg.name = msg_name
+    targets = np.linspace(0.0,-1.1, 200)
+    targets_el = np.linspace(0, 0.6, 200)
+    targets_wr = np.linspace(-1.5, -0.05, 200)
+    targets_head = np.linspace(-0.56, 0.56, 200)
```
Comment (missxa) [inline scripts/wrist_test.py]: in the test all scene, we should move all joints in the range of the limits randomly

File: scripts/wrist_test.py [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+    hand_shake(notest_lr)
+    # move_wrist(test_lr, 0)
+    move_wrist(test_lr, 1)
+    move_wrist(test_lr, 2)
+
+
+if __name__ == '__main__':
+
+    while publisher.get_num_connections() == 0:
+        rospy.loginfo("Waiting for subscriber to connect")
```
Comment (missxa) [inline scripts/wrist_test.py]: what subscriber?

**Review by missxa — CHANGES_REQUESTED** [review]

File: scripts/wrist_test.py (hunk 1 above)
Comment (thobotics) [reply scripts/wrist_test.py]: It is an abbreviation for left_or_right. Updated in new commit.

File: scripts/wrist_test.py (hunk 2 above)
Comment (thobotics) [reply scripts/wrist_test.py]: Removed

File: scripts/wrist_test.py (hunk 3 above)
Comment (thobotics) [reply scripts/wrist_test.py]: Done

File: scripts/wrist_test.py (hunk 4 above)
Comment (thobotics) [reply scripts/wrist_test.py]: Removed redundant logs and added more informative ones

File: scripts/wrist_test.py (hunk 5 above)
Comment (thobotics) [reply scripts/wrist_test.py]: Restructured the function by move_shoulder_elbow_wrist which provides users options to move different parts of an arm

File: scripts/wrist_test.py (hunk 6 above)
Comment (thobotics) [reply scripts/wrist_test.py]: Done

File: scripts/wrist_test.py (hunk 7 above)
Comment (thobotics) [reply scripts/wrist_test.py]: Done

File: scripts/wrist_test.py (hunk 8 above)
Comment (thobotics) [reply scripts/wrist_test.py]: Done

File: scripts/wrist_test.py (hunk 9 above)
Comment (thobotics) [reply scripts/wrist_test.py]: Changed the text

File: scripts/wrist_test.py [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+    for i in range(num_samples):
+        msg.position = list(targets[:, i])
+        msg.velocity = [0] * len(msg.name)
+        msg.effort = [0] * len(msg.name)
+        publisher.publish(msg)
+        rate.sleep()
+
+
+def move_wrist(left_or_right="right", axis=0, ranges=None):
+    ranges = [-1.5, -0.05] if not ranges else ranges
```
Comment (missxa) [inline scripts/wrist_test.py]:
```suggestion
def move_wrist(left_or_right="right", axis=0, ranges=[-1.5, -0.05]):

```

Comment (missxa) [reply scripts/wrist_test.py]: hm, it doesnt make sense to put these ranges by default. either pick a range that would work for every joint in roboy or leave it empty
also see my edit how to assign values to params

File: scripts/wrist_test.py (hunk 7 above)
Comment (missxa) [reply scripts/wrist_test.py]: i dont see `samples` in the list of parameters. where is it done?

File: scripts/wrist_test.py [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+                msg.position = [t, t]
+                publisher.publish(msg)
+                rospy.sleep(0.01)
+
+
+def move_all(move_type="sequentially", shoulder_range=None, elbow_range=None, wrist_range=None, head_range=None):
+    shoulder_range = [0.0, -1.1] if not shoulder_range else shoulder_range
+    elbow_range = [0.0, 0.6] if not elbow_range else elbow_range
+    wrist_range = [-1.5, -0.05] if not wrist_range else wrist_range
+    head_range = [-0.56, 0.56] if not head_range else head_range
```
Comment (missxa) [inline scripts/wrist_test.py]:
```suggestion
def move_all(move_type="sequentially", shoulder_range=[0.0, -1.1], elbow_range=[0.0, 0.6], wrist_range=[-1.5, -0.05], head_range= [-0.56, 0.56]):
```

File: scripts/wrist_test.py [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+
+    msg = JointState()
+    msg.name = ["shoulder_left_axis0", "elbow_left_axis0", "wrist_left_axis0", "wrist_left_axis1",
+                "shoulder_right_axis0", "elbow_right_axis0", "wrist_right_axis0", "wrist_right_axis1",
+                "head_axis0"]
+
+    if move_type == "sequentially":
```
Comment (missxa) [inline scripts/wrist_test.py]: strings are not good here, use enums for "randomly" "sequentially" etc or define in the beginning of the file
`SEQUENTIALLY = "sequentially"`

File: scripts/wrist_test.py [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+        targets = np.linspace(shoulder_range[0], shoulder_range[1], 200)
```
Comment (missxa) [inline scripts/wrist_test.py]: 200 - the last parameter is hardcoded everyone. it's bad style of coding. use a variable, and as mentioned in my comment above, it must be a parameter passed to the function

File: scripts/wrist_test.py [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+        publisher.publish(msg)
+        rate.sleep()
+
+
+def test_scene_1():
+    rospy.loginfo("########## SCENE 1 ##########")
+    rospy.loginfo("Description: Only move the wrist")
+    hard_reset()
+    rospy.loginfo("Move left wrist")
+    move_wrist("left", 2)
```
Comment (missxa) [inline scripts/wrist_test.py]: same here. dont use strings. they are subject to error. use enum or const strings definitions

File: scripts/wrist_test.py [hunk 15]
```diff
@@ -0,0 +1,245 @@
...
+                msg.effort = [0]
+                publisher.publish(msg)
+                rate.sleep()
+
+
+def move_shoulder_elbow_wrist(left_or_right="right", move_elbow=True, move_wrist=True,
+                              shoulder_range=None, elbow_range=None, wrist_range=None):
+    shoulder_range = [0.0, -1.1] if not shoulder_range else shoulder_range
+    elbow_range = [0.0, 0.6] if not elbow_range else elbow_range
+    wrist_range = [-1.5, -0.05] if not wrist_range else wrist_range
```
Comment (missxa) [inline scripts/wrist_test.py]: use default parameters correctly - see example in other comments

File: scripts/wrist_test.py [hunk 16]
```diff
@@ -0,0 +1,245 @@
...
+    msg.position = [0] * len(msg.name)
+    msg.velocity = [0] * len(msg.name)
+    msg.effort = [0] * len(msg.name)
+    publisher.publish(msg)
+
+    rospy.loginfo("Hard reset")
+    rospy.sleep(2.0)
+
+
+def hand_shake(left_or_right="right", ranges=None):
```
Comment (missxa) [inline scripts/wrist_test.py]: same here for parameter ranges

File: scripts/wrist_test.py [hunk 17]
```diff
@@ -0,0 +1,245 @@
+#!/usr/bin/env python
+
+import rospy
+from sensor_msgs.msg import JointState
+import numpy as np
+
+rospy.init_node("wrist_test")
+topic_root = '/roboy/pinky/'
```
Comment (missxa) [inline scripts/wrist_test.py:26]: unused variable

**Review by missxa — CHANGES_REQUESTED** [review]

File: scripts/wrist_test.py (hunk 17 above)
Comment (thobotics) [reply scripts/wrist_test.py:26]: Done

File: scripts/wrist_test.py (hunk 16 above)
Comment (thobotics) [reply scripts/wrist_test.py]: Done

File: scripts/wrist_test.py (hunk 15 above)
Comment (thobotics) [reply scripts/wrist_test.py]: Done

File: scripts/wrist_test.py (hunk 14 above)
Comment (thobotics) [reply scripts/wrist_test.py]: Done

File: scripts/wrist_test.py (hunk 13 above)
Comment (thobotics) [reply scripts/wrist_test.py]: Done

File: scripts/wrist_test.py (hunk 12 above)
Comment (thobotics) [reply scripts/wrist_test.py]: Done

File: scripts/wrist_test.py (hunk 10 above)
Comment (thobotics) [reply scripts/wrist_test.py]: New commit allows user to choose either manual (provide specific ranges and num_samples) or automatic testing (based on predefined scenarios)

File: scripts/wrist_test.py [hunk 18]
```diff
@@ -0,0 +1,298 @@
...
+    msg.name = ["shoulder_left_axis0", "elbow_left_axis0", "wrist_left_axis0", "wrist_left_axis1",
+                "shoulder_right_axis0", "elbow_right_axis0", "wrist_right_axis0", "wrist_right_axis1",
+                "head_axis0"]
+
+    if move_style == MoveStyle.SEQUENTIALLY:
+        targets = np.linspace(shoulder_range[0], shoulder_range[1], shoulder_samples)
+        targets_el = np.linspace(elbow_range[0], elbow_range[1], elbow_samples)
+        targets_wr = np.linspace(wrist_range[0], wrist_range[1], wrist_samples)
+        targets_head = np.linspace(head_range[0], head_range[1], head_samples)
+    elif move_style == MoveStyle.RANDOMLY:
```
Comment (missxa) [inline scripts/wrist_test.py:147]: check if simulated == true here and log a warning or error and ignore the joint targets

**Review by missxa — CHANGES_REQUESTED** [review]
almost there :)

File: scripts/wrist_test.py (hunk 18 above)
Comment (thobotics) [reply scripts/wrist_test.py:147]: I already put the checking for simulation in the test_scene_4 (line 246). In my opinion, the move_all function acts like an abstract that cover all move style, so I decided to let the caller (test_scene_4) make the decision (via is_simulated param) instead.

Comment (missxa) [reply scripts/wrist_test.py:147]: yes, i saw it, but you should have a check here, cause there might be other ppl using your script and your code must be idiot-proof, so right now technically I can still call this function with random movements with a physical robot.

Comment (thobotics) [reply scripts/wrist_test.py:147]: Updated

**Review by missxa — APPROVED** [review]

**Pull request merged**